src/libexpr/primops.cc

```diff
@@ -2472,6 +2472,93 @@ static RegisterPrimOp primop_mapAttrs({
     .fun = prim_mapAttrs,
 });
 
+static void prim_zipAttrsWith(EvalState & state, const Pos & pos, Value * * args, Value & v)
+{
+    // we will first count how many values are present for each given key.
+    // we then allocate a single attrset and pre-populate it with lists of
+    // appropriate sizes, stash the pointers to the list elements of each,
+    // and populate the lists. after that we replace the list in the every
+    // attribute with the merge function application. this way we need not
+    // use (slightly slower) temporary storage the GC does not know about.
+
+    std::map<Symbol, size_t> attrOccurences;
+    std::map<Symbol, Value * *> lists;
```

Comment:
`std::unordered_map` is probably faster. Also, it should be possible to merge these maps (`std::unordered_map<Symbol, std::tuple<size_t, Value * * >>`, where the `Value * *` is filled in in the second pass).

Reply:
`Symbol` isn't hashable yet, we'd have to add that. didn't do it because in our tests `unordered_map` was actually slower than `map`. (the sort at the end of the function is actually a leftover from those tests that snuck through)